<img src="https://weclouddata.s3.amazonaws.com/images/logos/wcd_logo_new_2.png"  width='15%'>  


<h1> Exercise Solution: LLM Evaluation - Basic Metrics</h1>
Developed by WeCloudData
<br></br>




In [1]:
!pip install evaluate --quiet

## 1. BLEU Score

A BLEU score ranges between 0 and 1 and measures how similar the machine-translated text is to a set of high quality reference translations:
- A value of 0 means that the output has no overlap with the reference hence the quality is low.
- A value of 1 means that there is perfect overlap with the reference data. So how can we apply this metric to evaluate the quality of llm’s output.


In [2]:
import evaluate
bleu = evaluate.load('bleu')
predictions = ["Transformers Transformers are fast plus efficient",
               "Good Morning", "I am waiting for new Transformers"]
references = [
              ["HuggingFace Transformers are quick, efficient and awesome",
               "Transformers are awesome because they are fast to execute"],
              ["Good Morning Transformers", "Morning Transformers"],
              ["People are eagerly waiting for new Transformer models",
               "People are very excited about new Transformers"]

]
results = bleu.compute(predictions=predictions,
                       references=references,
                       max_order = 2)
print(results)

The cache for model files in Transformers v4.22.0 has been updated. Migrating your old cache. This is a one-time only operation. You can interrupt this and resume the migration later on by calling `transformers.utils.move_cache()`.


0it [00:00, ?it/s]

/usr/local/lib/python3.12/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


{'bleu': 0.5037930378757725, 'precisions': [0.7142857142857143, 0.5454545454545454], 'brevity_penalty': 0.8071177470053892, 'length_ratio': 0.8235294117647058, 'translation_length': 14, 'reference_length': 17}


### BLEU Exercise

In [22]:
import transformers
print(transformers.__version__)

4.41.2


In [4]:
pip install -U transformers accelerate huggingface_hub

  Using cached transformers-5.8.1-py3-none-any.whl.metadata (33 kB)
  Using cached huggingface_hub-1.15.0-py3-none-any.whl.metadata (14 kB)
Using cached transformers-5.8.1-py3-none-any.whl (10.6 MB)
Using cached huggingface_hub-1.15.0-py3-none-any.whl (663 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.3/3.3 MB 50.3 MB/s eta 0:00:00
  Attempting uninstall: huggingface_hub
    Found existing installation: huggingface_hub 0.36.2
    Uninstalling huggingface_hub-0.36.2:
      Successfully uninstalled huggingface_hub-0.36.2
  Attempting uninstall: tokenizers
    Found existing installation: tokenizers 0.19.1
    Uninstalling tokenizers-0.19.1:
      Successfully uninstalled tokenizers-0.19.1
  Attempting uninstall: transformers
    Found existing installation: transformers 4.41.2
    Uninstalling transformers-4.41.2:
      Successfully uninstalled transformers-4.41.2


In [1]:
pip install transformers==4.41.2 accelerate huggingface_hub -U

  Using cached transformers-4.41.2-py3-none-any.whl.metadata (43 kB)
  Using cached huggingface_hub-0.36.2-py3-none-any.whl.metadata (15 kB)
  Using cached tokenizers-0.19.1-cp312-cp312-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (6.7 kB)
Using cached transformers-4.41.2-py3-none-any.whl (9.1 MB)
Using cached huggingface_hub-0.36.2-py3-none-any.whl (566 kB)
Using cached tokenizers-0.19.1-cp312-cp312-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (3.6 MB)
  Attempting uninstall: huggingface_hub
    Found existing installation: huggingface_hub 1.15.0
    Uninstalling huggingface_hub-1.15.0:
      Successfully uninstalled huggingface_hub-1.15.0
  Attempting uninstall: tokenizers
    Found existing installation: tokenizers 0.22.2
    Uninstalling tokenizers-0.22.2:
      Successfully uninstalled tokenizers-0.22.2
  Attempting uninstall: transformers
    Found existing installation: transformers 5.8.1
    Uninstalling transformers-5.8.1:
      Successfully uninstalled transforme

In [2]:
import transformers
print(transformers.__version__)

4.41.2


In [3]:
from transformers import pipeline, set_seed


set_seed(42)

def create_pipeline(
    task = 'summarization', model = 'facebook/bart-large-cnn', tokenizer = 'facebook/bart-large-cnn'
) -> pipeline:
    return pipeline(task=task, model=model, tokenizer=tokenizer)

summarizer = create_pipeline()

/usr/local/lib/python3.12/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


model.safetensors:   0%|          | 0.00/1.63G [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/363 [00:00<?, ?B/s]

vocab.json: 0.00B [00:00, ?B/s]

merges.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

In [4]:
!wget https://s3.amazonaws.com/weclouddata/datasets/llm/depression_information.txt

--2026-05-18 09:10:06--  https://s3.amazonaws.com/weclouddata/datasets/llm/depression_information.txt
Resolving s3.amazonaws.com (s3.amazonaws.com)... 16.15.228.119, 16.15.254.230, 16.15.212.21, ...
Connecting to s3.amazonaws.com (s3.amazonaws.com)|16.15.228.119|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 3440 (3.4K) [text/plain]
Saving to: ‘depression_information.txt’

depression_informat 100%[===================>]   3.36K  --.-KB/s    in 0s      

2026-05-18 09:10:06 (90.5 MB/s) - ‘depression_information.txt’ saved [3440/3440]



In [5]:
def read_information(
    file_name = 'depression_information.txt'
) -> str:
    return open('depression_information.txt', 'r').read()

data = read_information()

In [6]:
len(data)

3438

In [7]:
data

'Overview - Depression in adults\nDepression is more than simply feeling unhappy or fed up for a few days.\n\nMost people go through periods of feeling down, but when you\'re depressed you feel persistently sad for weeks or months, rather than just a few days.\n\nSome people think depression is trivial and not a genuine health condition. They\'re wrong – it is a real illness with real symptoms. Depression is not a sign of weakness or something you can "snap out of" by "pulling yourself together".\n\nThe good news is that with the right treatment and support, most people with depression can make a full recovery.\n\nThese pages are about depression in adults. Read about depression in children and young people.\n\nHow to tell if you have depression\nDepression affects people in different ways and can cause a wide variety of symptoms.\n\nThey range from lasting feelings of unhappiness and hopelessness, to losing interest in the things you used to enjoy and feeling very tearful. Many people

In [8]:
import evaluate

result = summarizer(data, max_length=500)
summarised_text = result[0]['summary_text']

bleu = evaluate.load('bleu')
scores = bleu.compute(predictions=[summarised_text], references=[data])

print(scores)

{'bleu': 0.00015729763123912478, 'precisions': [0.9696969696969697, 0.8923076923076924, 0.828125, 0.7619047619047619], 'brevity_penalty': 0.00018299337534830135, 'length_ratio': 0.10410094637223975, 'translation_length': 66, 'reference_length': 634}


In [9]:
len(result)

1

In [10]:
result

[{'summary_text': 'Depression affects people in different ways and can cause a wide variety of symptoms. Symptoms of depression range from mild to severe. At its mildest, you may simply feel persistently low in spirit. Severe depression can make you feel suicidal, that life is no longer worth living. Treatment for depression can involve a combination of lifestyle changes and medicine.'}]

In [11]:
random_text = 'Thisa poor score should produce '
random_score = bleu.compute(predictions=[random_text], references=[data])
random_score

{'bleu': 0.0,
 'precisions': [0.0, 0.0, 0.0, 0.0],
 'brevity_penalty': 2.3214224311468502e-55,
 'length_ratio': 0.007886435331230283,
 'translation_length': 5,
 'reference_length': 634}

## 2. Rouge Score


In [12]:
!pip install rouge_score --quiet

  Preparing metadata (setup.py) ... done


In [13]:
from rouge_score import rouge_scorer

scorer = rouge_scorer.RougeScorer(['rouge1', 'rougeL'], use_stemmer=True)
scores = scorer.score('The quick brown dog jumps over the lazy fox.',
                      'The quick brown fox jumps over the lazy dog.')
print(scores)

{'rouge1': Score(precision=1.0, recall=1.0, fmeasure=1.0), 'rougeL': Score(precision=0.7777777777777778, recall=0.7777777777777778, fmeasure=0.7777777777777778)}


### Rogue Score Exercise

In [14]:
from transformers import pipeline, set_seed

set_seed(42)

def create_pipeline(
    task = 'summarization', model = 'facebook/bart-large-cnn', tokenizer = 'facebook/bart-large-cnn'
) -> pipeline:
    return pipeline(task=task, model=model, tokenizer=tokenizer)

summarizer = create_pipeline()

In [15]:
def read_information(
    file_name = 'depression_information.txt'
) -> str:
    return open('depression_information.txt', 'r').read()

data = read_information()

In [16]:
import evaluate

result = summarizer(data, max_length=500)
summarised_text = result[0]['summary_text']

rouge = evaluate.load('rouge')
scores = rouge.compute(predictions=[summarised_text], references=[data])

print(scores)

{'rouge1': np.float64(0.18849840255591052), 'rouge2': np.float64(0.17307692307692307), 'rougeL': np.float64(0.18849840255591052), 'rougeLsum': np.float64(0.18849840255591052)}


In [17]:
summarised_text

'Depression affects people in different ways and can cause a wide variety of symptoms. Symptoms of depression range from mild to severe. At its mildest, you may simply feel persistently low in spirit. Severe depression can make you feel suicidal, that life is no longer worth living. Treatment for depression can involve a combination of lifestyle changes and medicine.'

## Meteor Score

In [18]:
!pip install nltk

In [19]:
from nltk.translate import meteor_score
from nltk import word_tokenize

import nltk

# Calculate the BLEU score
nltk.download('punkt_tab')
nltk.download('wordnet', download_dir='/usr/local/share/nltk_data')

reference = "The quick brown fox jumps over the lazy dog"
candidate = "The fast brown fox jumps over the lazy dog"
tokenized_reference = word_tokenize(reference)
tokenized_candidate = word_tokenize(candidate)
score = meteor_score.meteor_score([tokenized_reference], tokenized_candidate)
print(score)

[nltk_data] Downloading package punkt_tab to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt_tab.zip.
[nltk_data] Downloading package wordnet to
[nltk_data]     /usr/local/share/nltk_data...


0.9993141289437586


- METEOR scores range from 0 to 1, with higher scores indicating better
similarity between the candidate and reference sentences.
- In this case, the high METEOR score of approximately 0.9993 indicates that the candidate sentence is very similar to the reference sentence according to the METEOR metric. This suggests that the candidate sentence is almost identical to the reference sentence in terms of lexical overlap, word order, and linguistic quality.

### Meteor Score Exercise

## BERTScore

In [20]:
!pip install bert_score --quiet

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 61.1/61.1 kB 4.5 MB/s eta 0:00:00


In [21]:
import torch
from bert_score import score


# reference and generated texts
ref_text = data
gen_text = summarised_text

# compute Bert score
P, R, F1 = score([gen_text], [ref_text], lang="en", model_type="bert-base-uncased")

# print results
print(f"Bert score: P={P.item():.4f} R={R.item():.4f} F1={F1.item():.4f}")

tokenizer_config.json:   0%|          | 0.00/48.0 [00:00<?, ?B/s]

/usr/local/lib/python3.12/dist-packages/huggingface_hub/file_download.py:949: FutureWarning: `resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.
  warnings.warn(


config.json:   0%|          | 0.00/570 [00:00<?, ?B/s]

vocab.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

model.safetensors:   0%|          | 0.00/440M [00:00<?, ?B/s]

Bert score: P=0.8812 R=0.5436 F1=0.6724


BERTScore is a metric for evaluating the quality of generated text, which leverages pre-trained BERT models.

- **Precision (P):** Precision measures the proportion of correctly identified relevant words in the generated summary compared to the reference text. A precision score of 0.8812 indicates that approximately 88.12% of the words in the generated summary are relevant to the reference text.

- **Recall (R):** Recall measures the proportion of relevant words in the reference text that are correctly identified in the generated summary. A recall score of 0.5436 indicates that approximately 54.36% of the relevant words in the reference text are present in the generated summary.